server/utils/bookingUtils.ts

```diff
@@ -93,3 +93,119 @@ export const getLatestExtension = (booking: Booking) => {
     return latestTime > testTime ? latestExtension : testExtension
   }, booking.extensions?.[0])
 }
+
+export const deriveBookingHistory = (booking: Booking) => {
+  const extensions = [...booking.extensions].sort((a, b) => {
+    const dateA = DateFormats.convertIsoToDateObj(a.createdAt)
+    const dateB = DateFormats.convertIsoToDateObj(b.createdAt)
+
+    return dateA.getTime() - dateB.getTime()
+  })
+
+  const bookingWithSortedExensions = {
+    ...booking,
+    extensions,
+  }
+
+  const history = []
+
+  for (let previous = bookingWithSortedExensions; previous; previous = getPreviousBookingState(previous)) {
+    history.push({ booking: previous, updatedAt: getUpdatedAt(previous) })
+  }
+  history.reverse()
+
+  return history
+}
+
+const getUpdatedAt = (booking: Booking): string => {
+  if (booking.status === 'departed') {
+    return booking.departure.createdAt
+  }
+  if (booking.status === 'arrived') {
+    if (booking.extensions.length === 0) {
+      return booking.arrival.createdAt
+    }
+    return booking.extensions[booking.extensions.length - 1].createdAt
+  }
+  if (booking.status === 'cancelled') {
+    return booking.cancellation.createdAt
+  }
+  if (booking.status === 'confirmed') {
+    return booking.confirmation.createdAt
+  }
+  return booking.createdAt
+}
+
+const getPreviousBookingState = (booking: Booking): Booking => {
+  switch (booking.status) {
+    case 'departed':
+      return getPreviousDepartedBookingState(booking)
+    case 'arrived':
+      return getPreviousArrivedBookingState(booking)
+    case 'cancelled':
+      return getPreviousCancelledBookingState(booking)
+    case 'confirmed':
+      return getPreviousConfirmedBookingState(booking)
+    default:
+      return undefined
+  }
+}
+
+const getPreviousDepartedBookingState = (booking: Booking) => {
+  if (booking.extensions.length === 0) {
+    return {
+      ...booking,
+      status: 'arrived',
+      arrivalDate: booking.arrival.arrivalDate,
+      departureDate: booking.arrival.expectedDepartureDate,
+    } as Booking
+  }
+
+  return {
+    ...booking,
+    status: 'arrived',
+    arrivalDate: booking.arrival.arrivalDate,
+    departureDate: booking.extensions[booking.extensions.length - 1].newDepartureDate,
```

Comment:
is all the casting `as Booking` necessary? 
I'm guessing it is otherwise it wouldn't be here... would casting the status `as const` make it unneccessary?

Reply:
I've change it to
```
const getPreviousDepartedBookingState = (booking: Booking): Booking => {
```

to make it explicit without so much duplication